Java/core/src/main/java/com/amazon/randomcutforest/tree/AbstractCompactRandomCutTree.java

```diff
@@ -203,92 +215,175 @@ Point getPointFromPointReference(Integer pointIndex) {
-        nodeManager.delete(node);
+        delete(intValue(node));
+    }
+
+    protected void delete(int node) {
+        if (isLeaf(node)) {
+            leafStore.delete(getLeafIndexForTreeIndex(node));
+        } else {
+            nodeStore.delete(node);
+        }
     }
 
     @Override
     protected int getCutDimension(Integer node) {
-        return nodeManager.getCutDimension(node);
+        return nodeStore.getCutDimension(node);
     }
 
     @Override
     protected double getCutValue(Integer node) {
-        return nodeManager.getCutValue(node);
+        return nodeStore.getCutValue(node);
     }
 
     @Override
     protected Integer getLeftChild(Integer node) {
-        return nodeManager.getLeftChild(node);
+        return nodeStore.getLeftIndex(node);
     }
 
     @Override
     protected Integer getRightChild(Integer node) {
-        return nodeManager.getRightChild(node);
+        return nodeStore.getRightIndex(node);
     }
 
     @Override
-    protected void replaceNode(Integer parent, Integer child, Integer otherNode) {
-        nodeManager.replaceNode(parent, child, otherNode);
+    void replaceChild(Integer parent, Integer child, Integer otherNode) {
+        replaceChild(intValue(parent), intValue(child), intValue(otherNode));
+    }
+
+    protected void replaceChild(int parent, int child, int otherNode) {
+        nodeStore.replaceChild(parent, child, otherNode);
+        setParent(otherNode, parent);
     }
 
     @Override
     protected void replaceNodeBySibling(Integer grandParent, Integer parent, Integer node) {
-        nodeManager.replaceParentBySiblingOfNode(grandParent, parent, node);
+        replaceNodeBySibling(intValue(grandParent), intValue(parent), intValue(node));
+
+    }
+
+    protected void replaceNodeBySibling(int grandParent, int parent, int node) {
+        int sibling = nodeStore.getSibling(parent, node);
+        nodeStore.replaceChild(grandParent, parent, sibling);
+        setParent(sibling, grandParent);
     }
 
     @Override
     protected Integer addLeaf(Integer pointIndex) {
-        return nodeManager.addLeaf(null, pointIndex, 1);
+        return leafStore.addLeaf(NULL, pointIndex, 1);
     }
 
     @Override
     protected Integer addNode(Integer leftChild, Integer rightChild, int cutDimension, double cutValue, int mass) {
-        return nodeManager.addNode(null, leftChild, rightChild, cutDimension, cutValue, mass);
+        return addNode(intValue(leftChild), intValue(rightChild), cutDimension, cutValue, mass);
+    }
+
+    protected int addNode(int leftChild, int rightChild, int cutDimension, double cutValue, int mass) {
+        return nodeStore.addNode(NULL, leftChild, rightChild, cutDimension, cutValue, mass);
     }
 
     @Override
     protected void increaseMassOfAncestors(Integer mergedNode) {
-        nodeManager.increaseMassOfAncestors(mergedNode);
+        nodeStore.increaseMassOfAncestorsAndItself(getParent(mergedNode));
```

Comment:
Good catch, fixed. Note that by callint `mergedNode.intValue()` instead of `intValue(mergedNode)`, this will throw a null pointer exception if merged node is null, but I think that's reasonable behavior in this case.